larpandoracontent/LArTwoDReco/LArCosmicRay/TrackInEMShowerAlgorithm.cc

```diff
@@ -0,0 +1,877 @@
+    const ClusterList *const pClusterList, TwoDSlidingFitResultMap &microSlidingFitResultMap, TwoDSlidingFitResultMap &macroSlidingFitResultMap, ClusterVector &clusterVector) const
+{
+    // Determine the shower clusters which contain hits that belong to the main track
+    ClusterVector showerClustersToFragment;
+    for (const auto &entry : clusterToCaloHitListMap)
+    {
+        if ((entry.first != clusterAssociation.GetUpstreamCluster()) && (entry.first != clusterAssociation.GetDownstreamCluster()))
+            showerClustersToFragment.push_back(entry.first);
+    }
+
+    std::sort(showerClustersToFragment.begin(), showerClustersToFragment.end(), LArClusterHelper::SortByNHits);
+
+    ClusterList remnantClusterList;
+
+    const Cluster *const pMainTrackCluster(this->RemoveOffAxisHitsFromTrack(clusterAssociation.GetUpstreamCluster(), clusterAssociation.GetUpstreamMergePoint(),
+        true, clusterToCaloHitListMap, remnantClusterList, microSlidingFitResultMap, macroSlidingFitResultMap, clusterVector));
+    const Cluster *const pClusterToDelete(this->RemoveOffAxisHitsFromTrack(clusterAssociation.GetDownstreamCluster(), clusterAssociation.GetDownstreamMergePoint(),
+        false, clusterToCaloHitListMap, remnantClusterList, microSlidingFitResultMap, macroSlidingFitResultMap, clusterVector));
+
+    PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::MergeAndDeleteClusters(*this, pMainTrackCluster, pClusterToDelete));
+
+    for (const Cluster *const pShowerCluster : showerClustersToFragment)
+    {
+        const CaloHitList &caloHitsToMerge(clusterToCaloHitListMap.at(pShowerCluster));
+
+        this->UpdateForClusterDeletion(pShowerCluster, clusterVector, microSlidingFitResultMap, macroSlidingFitResultMap);
+        this->AddHitsToMainTrack(pShowerCluster, pMainTrackCluster, caloHitsToMerge, clusterAssociation, remnantClusterList);
+    }
+
+    this->ProcessRemnantClusters(remnantClusterList, pMainTrackCluster, pClusterList);
+    this->UpdateAfterMainTrackCreation(pMainTrackCluster, clusterVector, microSlidingFitResultMap, macroSlidingFitResultMap);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+const Cluster *TrackInEMShowerAlgorithm::RemoveOffAxisHitsFromTrack(const Cluster *const pCluster, const CartesianVector &splitPosition, const bool isUpstream,
+    const ClusterToCaloHitListMap &clusterToCaloHitListMap, ClusterList &remnantClusterList, TwoDSlidingFitResultMap &microSlidingFitResultMap,
+    TwoDSlidingFitResultMap &macroSlidingFitResultMap, ClusterVector &clusterVector) const
+{
+    const auto extrapolatedCaloHitIter(clusterToCaloHitListMap.find(pCluster));
+
+    if (extrapolatedCaloHitIter == clusterToCaloHitListMap.end())
+    {
+        this->UpdateForClusterDeletion(pCluster, clusterVector, microSlidingFitResultMap, macroSlidingFitResultMap);
```

Comment:
Is this the best place to be running this update for deletion, after checking in the clusterToCaloHitListMap? (The answer to this may well be "yes"!). Wondered whether there might be a better place. Good to have one last check through all the UpdateForClusterDeletion calls, as these are so tricky to make 100% robust.

Reply:
These are a real pain, and this comment made me realise that I wasn't handling something the way I wanted to. Essentially this statement is saying 'if the main track split cluster has no hits in the extrapolation vector skip the fragmentation process'. What I was trying to say was 'If there is no clustering errors then skip the fragmentation process'. In reality when you have a clustering error you nearly always find hits in the extrapolation vector but I can imagine that this isn't always true. In these cases, the current implementation means that the clustering error just wouldn't be fixed. So, I now have a 'more correct implementation' and have one cluster deletion update. Which I think is placed in the best place since it is that we can guarantee that the cluster exists.